contrib/gitlab.tpl

```diff
@@ -29,6 +29,10 @@
   {{- range . }}
   {{- $target := .Target }}
     {{- $image := $target | regexFind "[^\\s]+" }}
+    {{- $os := "Unknown" }}
+    {{- if contains "(" $target -}}
```

Comment:
I am worried that filepath might contain `(`:
https://github.com/aquasecurity/trivy/blob/56dbe1f6768fe67fbc1153b74fde0f83eaa1b281/pkg/scanner/langpkg/scan.go#L52

Can `Class` field be used to detect OS package vulnerabilities?
https://github.com/aquasecurity/trivy/blob/56dbe1f6768fe67fbc1153b74fde0f83eaa1b281/pkg/scanner/ospkg/scan.go#L40-L41

Reply:
My goal here was to replicate the data from GitLab's container scanning tool using Trivy directly. Within GitLab, the format is unique to container scanning (i.e. container images in a registry). In that use case, there's shouldn't be parentheses in the `.Target` value as they're invalid. If it's being used in other cases with file paths then parentheses could be an issue, but then spaces could also be an issue with line 31
```go
{{- $image := $target | regexFind "[^\\s]+" }}`
```

Reply:
On the `Class` question, GitLab actually displays the results, not just looking to identify OS packages, for example:
![image](https://github.com/user-attachments/assets/59cd67ae-2ced-48bc-9944-cdf97bd6d795)

Reply:
> If it's being used in other cases with file paths then parentheses could be an issue, but then spaces could also be an issue with line 31

Hm.. you are right.
We need to think about language-specific files from container.

I think we can check all targets to find `os` and `image` and use them for all targets.

Something like this:
```
  {{- $image := "Unknown" -}}
  {{- $os := "Unknown" -}}
  {{- range . }}
    {{- if eq .Class "os-pkgs" -}}
      {{- $target := .Target }}
        {{- $image = $target | regexFind "[^\\s]+" }}
        {{- $os = $target | splitList "(" | last | trimSuffix ")" }}
    {{- end }}
  {{- end }}
  "vulnerabilities": [
  {{- $t_first := true }}
  {{- range . }}
  {{- $target := .Target }}
    {{- range .Vulnerabilities -}}
```

wdyt?

Reply:
@DmitriyLewen I like it. I did a quick test and it is slightly different than what GitLab output, but I think it's more logical. By default they run with `--vuln-type os`, which I had forgotten, but after overriding they output `"operating_system": "Unknown"` for non-OS packages. There are some cases, for example I can think of some npm package examples, that have language-specific installations (installing binaries), so I think the OS is useful data.

At the moment there's one test for this template, which I had updated, but what are your thoughts on tests for the "Unknown" cases?

Reply:
> At the moment there's one test for this template, which I had updated, but what are your thoughts on tests for the "Unknown" cases?

Hm... You can add test for [testdata/fixtures/images/vulnimage.tar.gz](https://github.com/aquasecurity/trivy-test-images/pkgs/container/trivy-test-images/24420571?tag=vulnimage) (use `low` severity to skip a few vulns).

```
➜ trivy -q image  -f template -t @./contrib/gitlab.tpl  --input ./integration/testdata/fixtures/images/vulnimage.tar.gz --severity LOW | grep "operating_system" -A 1
        "operating_system": "alpine 3.7.1",
        "image": "./integration/testdata/fixtures/images/vulnimage.tar.gz"
--
        "operating_system": "alpine 3.7.1",
        "image": "./integration/testdata/fixtures/images/vulnimage.tar.gz"
```

Reply:
I was actually thinking the other direction - an image, maybe something `FROM scratch`, that would report `"operating_system": "Unknown"`, just to test the defaults. I'm not sure it's critical, but didn't see an image in the existing testcases that looked like it would work. I think the image/os values are already covered in the test for `integration/testdata/alpine-310.gitlab.golden`.

Reply:
I have provided `vulnimage.tar.gz` to show that vulnerabilities for language packages now contain the correct `operating_system` and `image` values.

Actual behavior:
```
➜  trivy -q image  -f template -t @./contrib/gitlab.tpl  --input ./integration/testdata/fixtures/images/vulnimage.tar.gz --severity LOW | grep "operating_system" -A 1
        "operating_system": "Unknown",
        "image": "./integration/testdata/fixtures/images/vulnimage.tar.gz"
--
        "operating_system": "Unknown",
        "image": "rust-app/Cargo.lock"

```

> that would report "operating_system": "Unknown", just to test the defaults

We can use `fs` mode (use one dir from [repo](https://github.com/aquasecurity/trivy/tree/main/integration/testdata/fixtures/repo) to check this. In this case OS is not detected.

Reply:
I added a test and the actions are all passing in my fork, so from my perspective I think this is ready.